feat: add weather easter egg

noa-santo opened the pull request:
### Features
- Added easter egg to aurora hero
- If activated weather data from user location gets fetched
- Weather gets displayed in aurora hero
- If weather is sunny the position of the sun is synced to real life

<details>
  <summary>Technical information</summary>

### Technical information
- Weather data and sunset/sunrise time gets fetched from api.weatherapi.com
- Resturctured aurora inside the hero component:
  1. Aurora types are now in a separate file (aurora-types.ts)
  2. Moved default flares from aurora hero to own component (default-flares.tsx)
  3. Moved AuroraContainer, AuroraBackground and AuroraForeground to aurora-components
- Added weather-easter-egg directory to aurora in hero component with:
  1. weather-api.ts to querry the API
  2. useWeatherEasterEgg custom Hook to use easter egg  (usage at [readme](#readme) )
  3. A component for every weather type (cloudy and slightly cloudy are one component)
  4. A custom hook (useWeather) to use the weather API to set a state to the current weather if shortcut is pressed 
  
</details>


### How to verify
1. [Open the website](https://satellytescommain21751-feataddweathereasteregg.gatsbyjs.io)
6. Press Shift + Control + Option (Mac) / Shift + Control + Alt (Windows)
7. Current weather should replace the hero


### Todo list
- [ ] add weather api token to 1password (name: GATSBY_WEATHER_API_KEY)

<details>
  <summary>Screenshots</summary>

### Screenshots
#### Easter egg disabled:
![image](https://github.com/satellytes/satellytes.com/assets/72336443/009a2f85-ce4f-46db-b6a4-dc0d7bfec90a)

#### Sunny weather:
![image](https://github.com/satellytes/satellytes.com/assets/72336443/c34f866d-40e2-4071-b0cb-3b366fde6bfd)

#### Slightly cloudy weather:
![image](https://github.com/satellytes/satellytes.com/assets/72336443/c26a8972-73a6-4580-9889-f29c54cafcf2)

#### Cloudy weather:
![image](https://github.com/satellytes/satellytes.com/assets/72336443/fd0afa73-991a-40ed-86c7-9744dab953ea)

#### Rainy weather:
![image](https://github.com/satellytes/satellytes.com/assets/72336443/965a89bd-60f6-447e-b0c5-fd0ac09422cf)

#### Snowy weather:
![image](https://github.com/satellytes/satellytes.com/assets/72336443/4f2f32b9-ce34-4422-ba4c-f64d97445275)

</details>

<!----><a name="readme"></a>
### Easter egg Readme:
[README.md](https://github.com/satellytes/satellytes.com/blob/833ef1e51ddd6cec64951af5dac8a546d009b12a/src/components/content/heroes/aurora/weather-easter-egg/README.md)

noa-santo committed:
feat: added weather-api.ts

noa-santo committed:
feat: added shortcut to toggleWeatherType; added WeatherType
fix: Bugfix in weather-api.ts (made getWeather a function)

noa-santo committed:
feat: added radial flare type for sun or snowflakes

noa-santo committed:
feat: changing bg according to weather

noa-santo committed:
feat: changing flares according to weather

noa-santo committed:
feat: finnished sunny flares

noa-santo committed:
fix: removed comment and removed unnecessary functions and dependency (from useEffect)

noa-santo committed:
feat: added debug url to weather aoi and added rain bg

noa-santo committed:
feat: added rainy bg

noa-santo committed:
feat: added rain when rainy

noa-santo committed:
feat: added snow

noa-santo committed:
feat: added cloud when cloudy

noa-santo committed:
fix: renamed Backgroud to Color

noa-santo committed:
fix: removed key property of Flares in flaresByWeather

noa-santo committed:
fix: readded normal api

noa-santo committed:
feat: added speed deviation to rain and snow

noa-santo committed:
fix: weather-api.ts now return always a vaild weather.

noa-santo committed:
feat: api now fetches local weather using the ip

noa-santo committed:
fix: added missing weather codes. removed unnecessary api call. weather-api.ts now returns WeatherType instead of String

noa-santo committed:
fix: removed commas in the top left corner

noa-santo committed:
fix: fixed all warnings

noa-santo committed:
fix: fixed background that was removed in last commit (now there are 2 weak warnings again)

noa-santo committed:
fix: getWeatherDescription now works as expected
feat: added test for getWeatherDescription
fix: aurora and weather types are now in aurora-types.ts

noa-santo committed:
feat: moved aurora components in extra file

noa-santo committed:
fix: aurora-hero.tsx now imports AuroraType from aurora/aurora-types.ts

noa-santo committed:
fixed typos

noa-santo committed:
feat: snowflakes now have a different size according to their speed => 3d effect

noa-santo committed:
feat: add weather-api (#10)

* feat: added weather-api.ts

* Update src/components/content/heroes/weather-api.ts

Co-authored-by: Jonathan Ostertag <jonathan@ostertage.de>

---------

Co-authored-by: Jonathan Ostertag <jonathan@ostertage.de>

noa-santo committed:
feat: add weather easter egg ui (#9)

* feat: added weather-api.ts

* feat: added shortcut to toggleWeatherType; added WeatherType
fix: Bugfix in weather-api.ts (made getWeather a function)

* feat: added radial flare type for sun or snowflakes

* feat: changing bg according to weather

* feat: changing flares according to weather

* feat: finnished sunny flares

* fix: removed comment and removed unnecessary functions and dependency (from useEffect)

* feat: added debug url to weather aoi and added rain bg

* feat: added rainy bg

* feat: added rain when rainy

* feat: added snow

* feat: added cloud when cloudy

* fix: renamed Backgroud to Color

* fix: removed key property of Flares in flaresByWeather

* fix: readded normal api

* feat: added speed deviation to rain and snow

* fix: weather-api.ts now return always a vaild weather.

* feat: api now fetches local weather using the ip

* fix: added missing weather codes. removed unnecessary api call. weather-api.ts now returns WeatherType instead of String

* fix: removed commas in the top left corner

* fix: fixed all warnings

* fix: fixed background that was removed in last commit (now there are 2 weak warnings again)

* fix: getWeatherDescription now works as expected
feat: added test for getWeatherDescription
fix: aurora and weather types are now in aurora-types.ts

* feat: moved aurora components in extra file

* fix: aurora-hero.tsx now imports AuroraType from aurora/aurora-types.ts

* fixed typos

noa-santo committed:
feat: snowflakes now move left and right while falling

noa-santo committed:
feat: clouds now move in random directions

noa-santo committed:
feat: better snow movement

noa-santo committed:
feat: weather code 1003 is now Sunny

noa-santo committed:
fix: renamed precipitationEffect to precipitation-effect
fix: renamed precipitationData to PrecipitationData

noa-santo committed:
fix: sway animation should now be random for every snowflake

noa-santo committed:
fix: Rainy Bg are now flares
feat: more customization in flare
fix: removed unnecessary cases
fix: made weather a required argument
fix: no negative positions

noa-santo committed:
feat: removed flares when cloudy
feat: made one CloudOverlay and one CloudShape with props instead of 6/2

noa-santo committed:
feat: more clouds

hofmann-felix committed:
refactor: move clouds into own component

noa-santo committed:
Merge branch 'feat/weather-flares-new-design'

# Conflicts:
#	src/components/content/heroes/aurora/aurora-components.tsx
#	src/components/content/heroes/aurora/aurora.tsx
#	src/components/content/heroes/aurora/weather-api.ts

noa-santo committed:
merge: feat: weather flares new design

noa-santo committed:
feat: weather flares new sun (no merge conflicts merge) (#16)

* feat: sun follow real sun

* fix: fixed formula to calculate height

* feat: sun now rotates

* feat: sun now looks better

* fix: tried to fix getSunlightPercentage (works every 5 refreshes)

* fix: tried to fix getSunlightPercentage (works every 4 refreshes)
feat: added tests

* Feat/weather flares better rain (#13)

* feat: rain has slightly different sizes now

* feat: drop size is now random

* feat: removed size differences and added blur to random drops

* feat: better blur

* feat: more contrast between drops and bg

* feat: better snow (#14)

* fix: sun is now at the right position (not 0 or infinity)

* fix: removed console logs
fix: timezone is now always correct

* fix: merged precipitationEffect.tsx into precipitation-effect.tsx

* fix: implemented feedback from PR

noa-santo committed:
feat: moved default flares to own file as component (#18)

noa-santo committed:
chore: Sun flares moved to sun component (#19)

* chore: Sun flares moved to sun component

* fix: added default flares

* fix: renamed AuroraSun to Sun and removed unnecessary <></>

noa-santo committed:
chore: rain restructure (#20)

* chore: moved rainy flares and rainy color and rain effect to own component

* fix: added default flares

noa-santo committed:
chore: moved snow to own component (#21)

noa-santo committed:
fix: removed unnecessary import

noa-santo committed:
chore: moved easter egg files in weather-easter-egg folder

noa-santo committed:
chore: removed magenta border of clouds (#22)

noa-santo committed:
Merge branch 'chore/weather-easter-egg-restructure'

noa-santo committed:
fix: snow and rain are visible now (#25)

* fix: snow and rain are visible now

* fix: added types to amount (snow, rain)

noa-santo committed:
chore: weather easter egg restructure (#26)

* feat: new custom hook: useWeather
feat: new component: WeatherEasterEgg
chore: moved dark default flares to default-flares

* chore: moved dark default flares back to aurora-components

noa-santo committed:
Feat/slightly cloudy (#27)

* feat: added weather type: slightly cloudy

* feat: added slightly cloudy to api

* feat: added slightly cloudy weather

* feat: more clouds (5 -> 10)

noa-santo committed:
chore: code refactor easter egg (#28)

* chore: added typing
chore: rgb -> hex
chore: Comments at complex formula

* chore: removed style from precipitation effect and added props

* chore: added more interfaces

ghost commented:
## :white_check_mark: satellytes.com-main deploy preview ready
* [Deploy preview](https://satellytescommain21751-feataddweathereasteregg.gatsbyjs.io)
* [Build logs](https://gatsbyjs.com/dashboard/f215130d-15e3-4176-9252-e13166fe48de/sites/674be368-adf0-4ac5-8fae-f9a4493adc51/builds/b8be88dc-c2f6-4b76-8931-b4aed7ddcae6/details) · **2m** build time
* Performance **76** :orange_circle: · Accessibility **97** :green_heart: · Best Practices **100** :green_heart: · SEO **86** :orange_circle: — [View full report](https://build-b8be88dc-c2f6-4b76-8931-b4aed7ddcae6.gtsb.io/reports/lighthouse/index.html)

milljoniaer committed:
Merge branch 'main' into feat/add-weather-easter-egg

noa-santo committed:
fix: added GATSBY_ prefix to env var

noa-santo committed:
Merge remote-tracking branch 'origin/feat/add-weather-easter-egg' into feat/add-weather-easter-egg

milljoniaer commented on src/components/content/heroes/aurora/aurora.tsx:
```diff
@@ -66,59 +30,14 @@ export const Aurora = ({ type, className }: AuroraProps) => {
-          rotation={80}
-          animationOffset={14}
-        />
+        <AuroraForeground>
```
Why do we wrap this twice in `<AuroraForeground>`?

milljoniaer commented on src/components/content/heroes/aurora/flare.ts:
```diff
@@ -63,15 +66,14 @@ function getFlareImage(type: FlareType) {
       background-image: url(${FlareTypeB});
     `;
-  } else {
+  } else if (type === FlareType.LIGHT) {
```
do we really need this?

milljoniaer commented on src/components/content/heroes/aurora/weather-easter-egg/clouds.tsx:
```diff
@@ -0,0 +1,107 @@
+  border-radius: 1000px;
+
+  position: absolute;
+  top: ${() => Math.floor(Math.random() * 100)}%;
```
Do we need `Math.floor()` here?

milljoniaer commented on src/components/content/heroes/aurora/weather-easter-egg/precipitation-effect.tsx:
```diff
@@ -0,0 +1,185 @@
+    50% 50% at 50% 50%,
+    rgba(255, 255, 255, ${(props) => props.transparency}) 0%,
+    rgba(255, 255, 255, ${(props) => props.transparency}) 20.83%,
+    rgba(255, 255, 255, ${(props) => props.transparency * 0.8}) 85.94%,
```
Do we need those uneven numbers? Does this add more value than 20% and 85%?

milljoniaer commented on src/components/content/heroes/aurora/weather-easter-egg/weather-easter-egg.tsx:
```diff
@@ -0,0 +1,11 @@
+  weather: WeatherType;
+}
+
+export const WeatherEasterEgg = ({ weather }: WeatherEasterEggProps) => {
```
This component seems unnecessary. Pls check if we can optimize this.

milljoniaer commented on src/components/content/heroes/aurora/weather-easter-egg/sun.tsx:
```diff
@@ -0,0 +1,129 @@
+    };
+
+    if (sunrise === 0 || sunset === 0)
+      fetchData().catch((e) => console.error(e));
```
`fetchData()` is already wrapped inside a try catch statement. You can get rid of the catch statement here.

milljoniaer commented on src/components/content/heroes/aurora/weather-easter-egg/sun.tsx:
```diff
@@ -0,0 +1,129 @@
+`;
+
+export const Sun = () => {
+  const [time, setTime] = useState(0);
```
`time` is the percentage of sunlight, isnt it? Please rename it.

milljoniaer commented on src/components/content/heroes/aurora/weather-easter-egg/sun.tsx:
```diff
@@ -0,0 +1,129 @@
+    return () => {
+      clearInterval(interval);
+    };
+  }, [sunset, sunrise, time]);
```
If you leave this array empty, the effect runs once at initial load of the component. Then you can get rid of the `if (sunrise === 0 || sunset === 0)` above.

milljoniaer commented on src/components/content/heroes/aurora/weather-easter-egg/sun.tsx:
```diff
@@ -0,0 +1,129 @@
+  position: absolute;
+`;
+
+const AuroraSunDiv = styled.div<{ timePercent: number }>`
```
What to do you think about adding a transition here, to let the sun fly into its correct position instead of jumping there (once the api request has responded)?

georgiee replied:
It would be nice to describe in a sentence what we can expect from this effect.

milljoniaer reviewed (commented):
Thank you for your work! Awesome that we get our first easter 🥚 !

I have added a few code style comments and general suggestions, please have a look

georgiee commented on src/components/content/heroes/aurora/aurora.tsx:
```diff
@@ -66,59 +30,14 @@ export const Aurora = ({ type, className }: AuroraProps) => {
-        />
+        <AuroraForeground>
+          <WeatherEasterEgg weather={weather} />
+        </AuroraForeground>
```
you have a double forgreound, I see no reason for that?

georgiee commented on src/components/content/heroes/aurora/aurora.tsx:
```diff
@@ -1,62 +1,26 @@
-  pointer-events: none;
-`;
+import { AuroraType } from './aurora-types';
+import {
+  AuroraForeground,
+  AuroraContainer,
+  AuroraBackground,
+} from './aurora-components';
```
Thanks for organizing the components 👍

georgiee commented on src/components/content/heroes/aurora/aurora-components.tsx:
```diff
@@ -0,0 +1,92 @@
+  pointer-events: none;
+`;
+
+export const flaresByWeather: { [key in WeatherType] } = {
```
that looks like heavily tied to the weather easter egg.  My expectation for the easter egg: Keep it as separated as possible.  You should maintain the flares inside the easter-egg folder and you can still export the "default flare" from this file.

georgiee commented on src/components/content/heroes/aurora/aurora-components.tsx:
```diff
@@ -0,0 +1,92 @@
+        return '#5A6FC1';
+      default:
+        return '#202840';
+    }
```
this switch creates a strong dependency to the weather/easter egg functionality. Isn't it possible to keep it outside? For example by accepting a prop for the background color and do the switch logic somewhere outside?

georgiee commented on src/components/content/heroes/aurora/aurora.tsx:
```diff
@@ -66,59 +30,14 @@ export const Aurora = ({ type, className }: AuroraProps) => {
   return (
     <AuroraContainer className={className}>
-      <AuroraBackground source={getSource(type)} />
+      <AuroraBackground source={getSource(type)} weather={weather} />
```
that weather you pass in is mostly for the color. Would be awesome if the background is not aware of any weather functionality. Less dependencies = easier maintenance.

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/weather-api.test.ts:
```diff
@@ -0,0 +1,134 @@
+
+jest.mock('axios');
+
+describe('getWeatherDescription', () => {
+  it.each([
+    [1000, WeatherType.Sunny],
+    [1003, WeatherType.SlightlyCloudy],
+    [1006, WeatherType.Cloudy],
+    [1009, WeatherType.Cloudy],
+    [1030, WeatherType.Cloudy],
+    [1063, WeatherType.Rainy],
```
Something went wrong. My location was detected as Frankfurt (Firefox magic?) and the bad weather their "rain shower" was interpreted as snow?

<img width="1486" alt="CleanShot 2023-07-19 at 14 43 40@2x" src="https://github.com/satellytes/satellytes.com/assets/1701755/21c1dd06-1d67-4c99-99e0-fa64b2052eeb">

noa-santo replied:
![Screenshot 2023-07-20 at 10 39 17](https://github.com/satellytes/satellytes.com/assets/72336443/d20150e5-7e41-44d3-a191-4e40b5fc02d1)
These weather codes were iterpreted as snowy but could also be rainy (changed non freezing drizzel to rain)

noa-santo committed:
chore: removed unnecessary second Foreground div

noa-santo committed:
chore: removed unnecessary if statement

noa-santo committed:
chore: removed unnecessary Math.floor()

noa-santo committed:
chore: removed unnecessary uneven numbers

noa-santo committed:
chore: removed unnecessary flaresByWeather in aurora-components and moved content to switch statement in WeatherEasterEgg

noa-santo committed:
chore: removed unnecessary catch (error already handled in weather-api)

noa-santo committed:
chore: removed unnecessary imports

noa-santo committed:
chore: renamed time to timePercent

noa-santo committed:
chore: removed unnecessary deps of useEffect

noa-santo committed:
chore: added animation to sun when timePercent changes suddenly (e.g. when first getting the data from api)

milljoniaer commented on src/components/content/heroes/aurora/weather-easter-egg/clouds.tsx:
```diff
@@ -0,0 +1,108 @@
+  position: absolute;
+
+  top: calc(${() => Math.floor(Math.random() * 100)}% - 200px);
+  left: calc(${() => Math.floor(Math.random() * 100)}% - 200px);
```
`Math.floor()` same as before 😄

milljoniaer commented on src/components/content/heroes/aurora/weather-easter-egg/clouds.tsx:
```diff
@@ -0,0 +1,108 @@
+
+export const CloudGroup = ({ type }: CloudGroupProps) => {
+  const amount = Math.floor(Math.random() * MAX_CLOUDS_PER_GROUP) + 1;
+  console.log(amount);
```
`console.log()` 🧹

milljoniaer commented on src/components/content/heroes/aurora/weather-easter-egg/clouds.tsx:
```diff
@@ -0,0 +1,108 @@
+};
+
+const Cloud = styled.div<{ type: CloudType }>`
+  --size: ${() => Math.floor(Math.random() * 400) + 100}px;
```
same here

milljoniaer commented on src/components/content/heroes/aurora/weather-easter-egg/sun.tsx:
```diff
@@ -0,0 +1,129 @@
+    return () => {
+      clearInterval(interval);
+    };
+  });
```
I didn't mean to remove the array, but keep an empty array. So it only runs at initial load

milljoniaer commented on src/components/content/heroes/aurora/weather-easter-egg/use-weather.ts:
```diff
@@ -0,0 +1,31 @@
+    return () => {
+      document.removeEventListener('keydown', handleKeyDown);
+    };
+  });
```
same here, empty array

noa-santo committed:
chore: clean up

noa-santo committed:
fix: useEffect in sun and useWeather is working again

noa-santo committed:
chore: added dark default flares to snowy and rainy weather

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/weather-api.test.ts:
```diff
@@ -0,0 +1,134 @@
+    [1225, WeatherType.Snowy],
+    [1237, WeatherType.Snowy],
+    [1240, WeatherType.Snowy],
+    [1243, WeatherType.Snowy],
```
Can you double check. My location was detected as Frankfurt. The weather was "rain shower" which got interpreted as snow. 

<img width="1486" alt="CleanShot 2023-07-19 at 14 43 40@2x" src="https://github.com/satellytes/satellytes.com/assets/1701755/21c1dd06-1d67-4c99-99e0-fa64b2052eeb">


That's the worst that can happen to a responsible developer: A test expectation suggesting a wrong reality.

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/weather-types.ts:
```diff
@@ -0,0 +1,8 @@
+export enum WeatherType {
```
I like enums ✨

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/weather-types.ts:
```diff
@@ -0,0 +1,8 @@
+  SlightlyCloudy = 'Slightly cloudy',
+  Rainy = 'Rainy',
+  Snowy = 'Snowy',
+  NotSet = 'Not set',
```
Is this an actual response or your default/fallback?

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/weather-easter-egg.tsx:
```diff
@@ -0,0 +1,33 @@
+  weather: WeatherType;
+}
+
+export const WeatherEasterEgg = ({ weather }: WeatherEasterEggProps) => {
```
That's the amount of separation I like. A single component that brings the feature and a single feature to remove or change if we want to update it.

There are some details of the easter egg currently bleeding into the aurora components but maybe it's fixable. I commented that separately at some other places.

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/weather-api.ts:
```diff
@@ -0,0 +1,96 @@
+  return WeatherType.NotSet;
+}
+
+export function convertTimeStringToTimestamp(timeString: string): number {
```
That works and the accompanying three tests are good too 🙌 

Your tests are now your safety rails ensuring that the outcome is correct which gives you a lot of confidence when refactoring. This means whenever you feel adventurous you can provide one of the following refactorings for the sake of trying out:

+ parse via regex (use regex101.com to make this fun)
+ involve date-fns's parse function as it's already part of our code base
+ Be smart with the Date class and its limited parsing capabilities

Nothing to do for you, just some thoughts

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/weather-api.test.ts:
```diff
@@ -0,0 +1,134 @@
+import axios, { AxiosResponse } from 'axios';
+import { getSunTime } from './weather-api';
+
+const API_KEY = process.env.WEATHER_API_KEY;
```
What's the differenct to `GATSBY_WEATHER_API_KEY` ?

noa-santo replied:
There is no difference. It had to be renamed so that the client can access it when it is deployed, so that the API can be used by the client when it is released.

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/use-weather.ts:
```diff
@@ -0,0 +1,30 @@
+import { useEffect, useState } from 'react';
+
+export function useWeather(): WeatherType {
+  const [weather, setWeather] = useState(WeatherType.NotSet);
```
It's good to use a defined value. Other folks would have used `null` to signal "not set". Such a value is called a sentinel value, as you use a unrelated value to signal "it's not set". Very good that you're using a dedicated enum

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/use-weather.ts:
```diff
@@ -0,0 +1,30 @@
+    document.addEventListener('keydown', handleKeyDown);
+
+    return () => {
+      document.removeEventListener('keydown', handleKeyDown);
```
thanks for the cleanup 👍

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/use-weather.ts:
```diff
@@ -0,0 +1,30 @@
+import { getWeather } from './weather-api';
+import { useEffect, useState } from 'react';
+
+export function useWeather(): WeatherType {
```
Have you thought about testing the hook? I have no experience testing hooks and the last few times I wanted to test them it was an unclear mess while it's often very useful and important as hooks can wrap quite some logic.

Any thoughts about that to share?

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/weather-api.ts:
```diff
@@ -0,0 +1,96 @@
+  sunsetTime: number;
+}
+
+export const getWeather = async (): Promise<WeatherType> => {
```
It's okay to remove the type. Typescript will determine it correctly.
Remove and check your IDE which will show the correct type;

<img width="753" alt="image" src="https://github.com/satellytes/satellytes.com/assets/1701755/585b699c-25de-4fbe-b660-3e0ad3e8d48b">

That's something many people do wrong because they think they have to declare it while they don't. Other folks just prefer to be explicit but I don't see any benefit in "over declaring" because pure functions are easily readable including the involved type.

georgiee commented on src/components/content/heroes/aurora/aurora-animation.ts:
```diff
@@ -1,8 +1,8 @@
+ * The `FRAMES` content has been created manually by running `transformMapToCssKeyframes` on the set of relative movement (0,1 etc.) from the `animation-tool.ts` file.
  * The map is roughly this pattern. I drew the ascii pattern and then manually transcribed it
- * as I didn't want to spend time processing it although it's fun, but it's distracting fun.
+ * as I didn't want to spend time processing it, although it's fun, but it's distracting fun.
```
thanks for reading this comment from me and taking your time to fix typos/commas 🙌

georgiee commented on src/components/content/heroes/aurora/aurora-components.ts:
```diff
@@ -0,0 +1,54 @@
+
+export const AuroraBackground = styled.div<AuroraBackgroundProps>`
+  background: ${(props) => {
+    switch (props.weather) {
```
I commented on this where it's used already. You pass in the weather to derive a color here right? You did it the simple way, pass in the weather and determine the color.

But if you use something like that on the outside;

```
<AuroraBackground background={getWeatherbackgroundColor()}
```

you remove the dependency of the weather/easter egg 🙌

--

I recognize that you're doing a little bit more when there is no weather but maybe it's also correct to say if there is no color I need a default special background?

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/sun.tsx:
```diff
@@ -0,0 +1,129 @@
+  position: absolute;
+`;
+
+export const Sun = () => {
```
That's such a good, subtle effect with the sun position included ☀️

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/sun.tsx:
```diff
@@ -0,0 +1,129 @@
+    }, 10000);
+
+    return () => {
+      clearInterval(interval);
```
thanks for the cleanup, easy to forget, important to have 👍

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/sun.tsx:
```diff
@@ -0,0 +1,129 @@
+    ${(props) =>
+        (1 / 350) *
+          -Math.pow(
+            (props.timePercent + (props.timePercent - 50) / 2) * 2 - 100, // parabola formula to get the sun reflection to move in a parabola
```
👍 
a reference of the formula will help future folks to understand it.
Plus I would create a pure function at the top of the file doing the calculatios. Make the css readable again.

Something like this:
```
bottom: calc(
     ${({timePercent} ) => parabola(timePercent )} 
```

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/sun.tsx:
```diff
@@ -0,0 +1,129 @@
+  height: 539px;
+  left: calc(
+    ${(props) => props.timePercent}% +
+      ${(props) => (props.timePercent - 50) / 2}% - 274.75px
```
`274.75px` is a super specific magic number. You got that by tinkering around?
What's the meaning? An offset? A center point? A variable name can help understanding it:

`${(props) => (props.timePercent - 50) / 2}% - ${centerX}` or similar

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/sun.test.ts:
```diff
@@ -0,0 +1,76 @@
+import { getSunlightPercentage } from './sun-percentage-calculator';
+
+describe('getSunlightPercentage', () => {
```
Thanks for testing that important part of your sun logic 🙌

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/sun-percentage-calculator.ts:
```diff
@@ -0,0 +1,21 @@
+  sunriseTime: number,
+  sunsetTime: number,
+) => {
+  const time = +new Date();
```
If your function accepts a parameter `time` you would gain a truly pure function that you can test without mocking the system time in your specs as you can control what to pass in. That's not crucial but something I absolutely recommend to change to make your code more understandable and easier to test 🙏

```
getSunlightPercentage(dateA, dateB, currentTime)
```

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/snow.tsx:
```diff
@@ -0,0 +1,33 @@
+  amount: number;
+}
+
+export const Snow: FC<SnowProps> = ({ amount }) => {
```
Seeing `React.FC` here I want to let you know that it's heavily discussed if you should use it or not.

Take this article and then start digging:
https://juhanajauhiainen.com/posts/should-you-use-reactfc-to-type-your-components

I'm still always torn, that's why it's more a FYI

georgiee replied:
Love it 🤩
![CleanShot 2023-07-20 at 09 27 30](https://github.com/satellytes/satellytes.com/assets/1701755/2ecdbd78-525d-4b29-8adf-c3cc4756ae25)

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/snow.tsx:
```diff
@@ -0,0 +1,33 @@
+        dropCount={amount}
+        speed={10}
+        type={PrecipitationType.Snow}
+        speedDeviation={3}
```
Nice idea to generalize the precipation effect

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/rain.tsx:
```diff
@@ -0,0 +1,58 @@
+  amount: number;
+}
+
+export const Rain: FC<RainProps> = ({ amount }) => {
```
My favourite effect. Only the drops splashing at the ground are missing to be super awesome and not only awesome 😁
![CleanShot 2023-07-20 at 09 27 09](https://github.com/satellytes/satellytes.com/assets/1701755/ba6a49c6-8582-493e-bd6a-8f58be7c9cdd)

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/rain.tsx:
```diff
@@ -0,0 +1,58 @@
+        dropCount={amount}
+        speed={1}
+        type={PrecipitationType.Rain}
+        speedDeviation={0.1}
```
I understand this param, but the higher the value of derviation was when I tinkered around with it the lower the speed got. Then I realized that this is not the derivation but the speed being wrong. You forward it directly into the animation as the duration which gets you lower speeds for higher numbers. That's not what I expect. Good enough for here but something to think about 🙏

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/precipitation-effect.tsx:
```diff
@@ -0,0 +1,185 @@
+  Snow = 'snow',
+}
+
+const PrecipitationEffect = ({ dropCount, speed, type, speedDeviation }) => {
```
I appreciate your generalization efforts to reuse this for rain & snow. 

Followggn some thoughts about the level of abstraction you chose. Actually nothing to change but maybe something to think about for you 👍

This function looks like you went the generalization path to early (kind of a premature optimization) as this effect looks more like a bundle of two effects that share some parameters and the little code around "generateDrops". 

I see a function with two big conditions and few more. Your code would be much easier to read and more importantly to experiment with and to derive other experiments if you have a dedicated RainEffect and SnowEffect instead of merging both into a single effect. Those effects could even include their own copy of `generateDrop` for the beginning. Doesn't actually matter and simple to share once you reach that point.

georgiee commented on src/components/content/heroes/aurora/aurora-components.ts:
```diff
@@ -0,0 +1,54 @@
+      case WeatherType.Rainy:
+        return '#76809b';
+      case WeatherType.SlightlyCloudy:
+        return '#5A6FC1';
```
That background makes it super hard to read for slightly cloudy. The sun one is much better to read. It's an easter egg and the visual is more important, but it's something that popped my eye.

<img width="576" alt="image" src="https://github.com/satellytes/satellytes.com/assets/1701755/be8366e7-e34c-4280-a591-6d02d9b40341">

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/clouds.tsx:
```diff
@@ -0,0 +1,107 @@
+    infinite;
+`;
+
+export const Clouds = ({ amount, type }: CloudsProps) => {
```
cool and flexible component ✨ the cloudy one really looks depressingly cloudy 😅
<img width="1234" alt="image" src="https://github.com/satellytes/satellytes.com/assets/1701755/3e5fda83-9500-4bd0-8352-a846e063771b">

slightly cloudy is bad for readability (while all other effects work) and it doesn't look very convincing.
<img width="1249" alt="image" src="https://github.com/satellytes/satellytes.com/assets/1701755/d82aa22a-1b73-471e-a34e-521e6dad2aef">

georgiee reviewed (approved):
Thank you for that awesome easter egg ✨ First of all: This change is ready to merge instantly which is a great achievement🙌 

I've commented a lot to get you some opportunities to think about your code. 

And there is one thing that you should fix before merging but it's not holding back this PR as you can deliver it after merging it and the actual easter egg works and it's not taking effect without knowing the keys anyway.

## Most Important Changes to deliver afterwards (if you have time left)
1. The png files are not optimized yet. Please make them much smaller. Right now, `sun-reflection.png` is `550k` or `sun-shine` has `450k`. That's not acceptable.

2. Provide a `README.md` in the easter egg folder. Explain how it works, what APIs are involved, what people can expect from the easter egg, maybe include images/videos and additional ideas for future folks to expand this easter egg if they like. You can also describe your motivation, that it was part of your internship etc. That would be a perfect wrap up

3. Remove the easter egg as a dependency from the aurora background to shrinken the dependency surface.

## Other things that I commented already
+ Make the percentage calculation for the sun pure by removing the +new Date(). That is a simple but powerful change and it will show you why people love pure functions 💛

## Ideas
While reading your code, especially where you mount the easter egg I thought about developer experience, a more pleasant API.

You could deliver the hook + component already which is nice already! But wouldn't it be possible to already deliver a prefilled component to have less of the wiring we have now? 

Something like this: (almost pseudo code)


```

export const Aurora  () => {
  const {weatherBackground, WeatherComponent} = useWeatherEasterEgg();

  return (
    <AuroraContainer className={className}>
      <AuroraBackground background={weatherBackground} />
      <AuroraForeground>
        <WeatherComponent/>
      </AuroraForeground>
    </AuroraContainer>
  );
};


```

You could even include some configuration for the keys or some way to override specific weather code to provide let's say a lightning effect or some [refraction effect](https://sciencedemonstrations.fas.harvard.edu/presentations/hot-road-mirage) when it's super super hot.

```
useWeatherEasterEgg({
  key: 'Ctrl + Cmd + G'
   codes: {
     1234: LightningEffect,
     1234: HotAirRefractionEffect
   }
});
```

georgiee commented:
I see the token todo for 1Password. How do we ensure the validity of the token in the future. Do we need to create an account ourselves ideally? Or can you create tokens without an account by chance?

noa-santo committed:
fix: better weather code interpretation

noa-santo committed:
feat: added test for useWeather

noa-santo committed:
chore: removed unnecessary type declaration

noa-santo committed:
chore: made sun divs more readable

noa-santo committed:
chore: made getSunlightPercentage a pure function

noa-santo committed:
chore: removed React.FC from snow and rain

noa-santo committed:
feat: added splashes to rain

noa-santo committed:
chore: renamed speed to fallDuration

noa-santo commented:
> I see the token todo for 1Password. How do we ensure the validity of the token in the future. Do we need to create an account ourselves ideally? Or can you create tokens without an account by chance?

The account for the api already uses technical.user@satellytes.com as email

noa-santo committed:
chore: removed aurora-components dependency on weather

noa-santo committed:
chore: added comment to sun

noa-santo committed:
feat: now the weather easter egg can be used with userWeatherEasterEgg()

noa-santo committed:
fix: added transition to sun reflection

noa-santo committed:
feat: useWeatherEasterEgg now supports custom shortcuts

noa-santo committed:
feat: useWeatherEasterEgg() now supports weather code overwrites

noa-santo committed:
feat: useWeatherEasterEgg() now supports new weather types

noa-santo committed:
feat: added README.md to weather-easter-egg

noa-santo committed:
fix: better slightly cloudy weather

noa-santo committed:
feat: better README.md

noa-santo committed:
fix: compressed sun assets (1 MB less total)

noa-santo commented:
**Changes:**

1. Improved weather code interpretation.
2. Added tests for useWeather.
3. Removed unnecessary type declaration.
4. Improved readability of sun divs.
5. Refactored getSunlightPercentage to be a pure function.
6. Removed React.FC from snow and rain components.
7. Added splashes to rain.
8. Renamed "speed" to "fallDuration" for clarity.
9. Enabled using weather easter egg with userWeatherEasterEgg().
10. Added transition to sun reflection.
11. Added support for custom shortcuts in useWeatherEasterEgg().
12. Added support for weather code overwrites in useWeatherEasterEgg().
13. Added support for new weather types in useWeatherEasterEgg().
14. Created README.md for weather-easter-egg.
15. Improved slightly cloudy weather.
16. Enhanced README.md content.
17. Compressed sun assets, reducing total size by 1 MB.

georgiee commented on src/components/content/heroes/aurora/aurora-components.ts:
```diff
@@ -0,0 +1,45 @@
+
+export const AuroraBackground = styled.div<AuroraBackgroundProps>`
+  ${(props) =>
+    props.overwriteBackground !== undefined
```
awesome, thanks for accepting and implementing my proposal regarding removing dependencies

georgiee commented on src/components/content/heroes/aurora/aurora.tsx:
```diff
@@ -66,59 +30,22 @@ export const Aurora = ({ type, className }: AuroraProps) => {
-          rotation={80}
-          animationOffset={14}
-        />
+        {isWeatherEasterEggEnabled ? (
+          WeatherComponent
+        ) : (
+          <>
+            <DefaultFlaresDark />
+            <DefaultFlares />
+          </>
+        )}
```
superb having a self contained weather component now 🙌

georgiee commented on src/components/content/heroes/aurora/aurora.tsx:
```diff
@@ -1,62 +1,26 @@
 
 export const Aurora = ({ type, className }: AuroraProps) => {
-  function getSource(type?: AuroraType) {
+  const { weatherBackground, WeatherComponent, isWeatherEasterEggEnabled } =
+    useWeatherEasterEgg({});
```
I love that hook DX

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/README.md:
```diff
@@ -0,0 +1,113 @@
+# Weather Easter Egg
```
Thank you for taking the idea of writing a README.md and taking it into even higher levels than expected. It's now a great example how to document a specific feature ✨

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/precipitation-effect.tsx:
```diff
@@ -0,0 +1,258 @@
+      const newDrops = Array.from({ length: dropCount }, (_, index) => ({
+        id: index,
+        color: '#84abd0',
+        fallDuration:
```
much better name than `speed` 👍

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/sun.tsx:
```diff
@@ -0,0 +1,141 @@
+  return (1 / 125) * (-(timePercent - 50) * (timePercent - 50)) + 60; // parabola formula to get the sun to move in a parabola
+};
+
+const getSunReflectionYPosition = (timePercent: number) => {
```
separation of concerncs. now css looks much better to read

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/use-weather-easter-egg.tsx:
```diff
@@ -0,0 +1,28 @@
+  const weather: WeatherType = useWeather({ key, codes });
+  const isWeatherEasterEggEnabled = weather !== WeatherType.NotSet;
+  const WeatherComponent = (
+    <WeatherEasterEgg weather={weather} customWeather={customWeather} />
```
👍

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/use-weather.test.ts:
```diff
@@ -0,0 +1,84 @@
+import { renderHook, act } from '@testing-library/react';
```
Thank you for educating me. From my "actually I don't know how to test hooks, it's not part of the react docs as far as I know" you found the great testing library and their approach of testing hooks 🤩👍

noa-santo committed:
feat: when Easter egg is active current weather can be overwritten with the number keys

georgiee commented on src/components/content/heroes/aurora/weather-easter-egg/weather-api.ts:
```diff
@@ -0,0 +1,112 @@
+  conditionCode: number,
+  customCodes: { [key: string]: string } | undefined,
+): WeatherType {
+  if (customCodes) {
+    return (
+      Object.keys(customCodes)
+        .map((range) => {
+          const customType = customCodes[range];
+          const [start, end] = range.split('-').map(Number);
+          if (start === conditionCode) {
+            return customType as WeatherType;
+          }
+          if (conditionCode >= start && conditionCode <= end) {
+            return customType as WeatherType;
+          }
+        })
+        .find((weatherType) => weatherType !== undefined) || WeatherType.NotSet
+    );
+  }
```
I did not expect that you make this possible 👏

noa-santo commented:
When easter egg is active the current weather can be overwritten with the number keys:
1: Sunny
2: Cloudy
3: Slightly Cloudy
4: Rainy
5: Snowy

georgiee reviewed (approved):
Thank you so much for all the changes after my review. Much more than expected and make this feature even more solid and appealing. Thanks!

Let's merge already 🤩✨  Any other feature that might sitting in the pipeline should be delivered through a separate PR because the review gets more difficult as the size and functionality of the PR was big from the beginning.
